### PR TITLE
DEIMOS mosaic

### PR DESCRIPTION
**Pull request by debora-pe**
Constructed mosaic for DEIMOS. Four mosaics total (one per each blue-red detectors pair). The blue detectors are the only ones to be transformed.

Documentation added.

Related Dev Suite PR: https://github.com/pypeit/PypeIt-development-suite/pull/194 (please consider reviewing that too)

Unrelated changes:
- several fixes to manual extraction
- fix bug with  `reduce_bpm`, which was updated in `objFind`, but not propagated to `Extract`
- fix bug in `chk_noise_2dspec`
- address issue #1359 

UNIT TESTS PASS:
`====================== 326 passed in 1300.42s (0:21:40) ======================`

### CODE REVIEW

File: doc/deimos.rst
Comment (profxj) [inline doc/deimos.rst]: typo?  I think it should read  "only i[f] the [] mosaic reduction"..

File: pypeit/manual_extract.py
Comment (profxj) [inline pypeit/manual_extract.py:77]: nice.  thanks for adding this!

File: pypeit/par/pypeitpar.py
Comment (profxj) [inline pypeit/par/pypeitpar.py:3714]: do we have a test of this in the DevSuite?  

If not, can we?

File: pypeit/par/pypeitpar.py
Comment (profxj) [inline pypeit/par/pypeitpar.py:2162]: Do we have a test of this in the DevSuite?  

If not, can we?

File: pypeit/scripts/chk_noise_2dspec.py
Comment (profxj) [inline pypeit/scripts/chk_noise_2dspec.py:111]: please add a short doc string

File: pypeit/spectrographs/gemini_gmos.py
Comment (profxj) [inline pypeit/spectrographs/gemini_gmos.py]: maybe put this in the `def` 

Is an easier spot to find (and then you can comment it there)

File: pypeit/spectrographs/keck_deimos.py
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py:342]: Is it possible to know here whether we are in mosaic mode and act accordingly?

same for `sigdetect` below..

File: pypeit/spectrographs/keck_deimos.py
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py:674]: I think this might have been used when trying to generate an image of all 8 detectors.  But can't quire recall.

File: pypeit/spectrographs/keck_deimos.py
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py]: As with GMOS, I suggest we add this to the `def` call.

And presumably need to update the method in `spectorgraph.py`

File: pypeit/spectrographs/keck_deimos.py
Comment (profxj) [inline pypeit/spectrographs/keck_deimos.py:1435]: Add a brief doc on what `mosaic` does

**Review by profxj — CHANGES_REQUESTED** [review]
Great stuff!

Only a few minor requests from me.

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py:52]: I assume these are the parameters when you're interpolating the red chip only?  I would add a comment here about this and indicate that it's useful to keep these around for now.

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py:233]: Does it need to be this large?  I think it only needs to be large enough that the trace "following" algorithm can span the chip gap.  I ask because this might lead to some spurious traces being connected, but I'm not sure.

Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py:233]: Does it need to be this large?  I think it only needs to be large enough that the trace "following" algorithm can span the chip gap.  I ask because this might lead to some spurious traces being connected, but I'm not sure.

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [reply pypeit/spectrographs/keck_deimos.py:342]: Good question.  It might be as simple as checking if there is a tuple in the detnum.

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [reply pypeit/spectrographs/keck_deimos.py:674]: The default `detnum` in `PypeItPar` is None, but I agree that this may never be the case now for DEIMOS.

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py:680]: Same thing here.

Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py:680]: Same thing here.

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py]: Does this mean `_ccdnum` and `_det` are identical?  Could you just set `_ccdnum` directly: `nimg, _ccdnum = self.validate_det(ccdnum)`?

Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py]: Does this mean `_ccdnum` and `_det` are identical?  Could you just set `_ccdnum` directly: `nimg, _ccdnum = self.validate_det(ccdnum)`?

File: pypeit/spectrographs/keck_deimos.py
Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py]: I would use the `pypeit.images.mosaic.Mosaic.get_name` method to set the mosaic name.

Comment (kbwestfall) [inline pypeit/spectrographs/keck_deimos.py]: I would use the `pypeit.images.mosaic.Mosaic.get_name` method to set the mosaic name.

**Review by kbwestfall — APPROVED** [review]
Great, @deborape !  I fixed some of the minor typos directly, and then I had a couple of minor suggestions.  I'm approving now though.

**Comment by kbwestfall** [comment]
Apologies.  For some reason, github made repeats of a bunch of my comments.  Let me know if you find anything confusing!

File: pypeit/par/pypeitpar.py
Comment (debora-pe) [reply pypeit/par/pypeitpar.py:2162]: Yes, we do. All the DEIMOS reductions include `detnum` in the pypeit file

File: pypeit/scripts/chk_noise_2dspec.py
Comment (debora-pe) [reply pypeit/scripts/chk_noise_2dspec.py:111]: Done

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py:52]: Done

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py:233]: The gap between the blue and red traces often was much larger than the chip gap. I tried different value, but this high value was giving the insurance to connect the traces.

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py:342]: Mosaic is the default. Only a different detnum added by the user in the pypeit file would make the reduction non-mosaic.
`config_specific_par` is run before making the merge with the parameters added by the user in pypeit file.  So at this point in the code, I cannot know whether is a mosaic or not.

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py:674]: Ok, I'll leave it, just in case.

File: pypeit/spectrographs/gemini_gmos.py
Comment (debora-pe) [reply pypeit/spectrographs/gemini_gmos.py]: Done

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py]: Done

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py:1435]: Done

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py]: I used `spectrograph.get_det_name`

File: pypeit/spectrographs/keck_deimos.py
Comment (debora-pe) [reply pypeit/spectrographs/keck_deimos.py]: Yes, that's right! Changed, thanks.

File: pypeit/par/pypeitpar.py
Comment (debora-pe) [reply pypeit/par/pypeitpar.py:3714]: Added

**Comment by debora-pe** [comment]
Dev Suite passes (There was a problem with one dataset, but resolved in the Dev Suite)
```
--- PYPEIT DEVELOPMENT SUITE FAILED 1/139 TESTS (Masters ignored) ---
Failed tests:
    keck_deimos/830G_M_9000_dither pypeit_coadd_2dspec
Skipped tests:
Testing Started at 2022-04-18T17:44:02.538756
Testing Completed at 2022-04-18T23:13:19.346728
Total Time: 5:29:16.807972
```


```
Running tests on the following instruments:
    keck_deimos

Reducing data from keck_deimos for the following setups:
    830G_M_9000_dither

 0 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_9000_dither pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_9000_dither pypeit (without masters)
 1 passed/ 0 failed/ 0 skipped STARTED keck_deimos/830G_M_9000_dither pypeit_coadd_2dspec
 2 passed/ 0 failed/ 0 skipped PASSED  keck_deimos/830G_M_9000_dither pypeit_coadd_2dspec

--- PYPEIT DEVELOPMENT SUITE PASSED 2/2 TESTS (Masters ignored) ---
Testing Started at 2022-04-19T14:16:07.631757
Testing Completed at 2022-04-19T15:54:44.587929
Total Time: 1:38:36.956172
```

**Review by profxj — APPROVED** [review]
Great stuff!

Approved.

**Comment by profxj** [comment]
Merge when you wish.

**Pull request merged**
